lib/net.ts

```diff
@@ -10,49 +10,57 @@ import * as http from 'http';
 import * as url from 'url';
 import * as Streams from './streams';
 
+export type PostData = string | {[key: string]: string | number};
+export interface NetRequestOptions extends https.RequestOptions {
+	body?: PostData;
+}
+
 export class NetRequest {
 	uri: string;
 	statusCode?: number;
 	constructor(uri: string) {
 		this.uri = uri;
 		this.statusCode = undefined;
 	}
-	/**
-	 * Makes a basic http/https request to the URI.
-	 * Returns the response data.
-	 * @param opts request opts - headers, etc.
-	 */
-	async get(opts?: AnyObject): Promise<string | null> {
-		if (Config.noNetRequests) throw new Error(`Net requests are disabled.`);
-		const protocol = url.parse(this.uri).protocol as string;
-		const net = protocol.includes('https:') ? https : http;
-		return new Promise((resolve, reject) => {
-			const req = net.get(opts ? opts : this.uri, res => {
-				res.setEncoding('utf-8');
-				this.statusCode = res.statusCode;
-				void Streams.readAll(res).then(buffer => resolve(buffer));
-			});
-			req.on('error', err => {
-				reject(err);
-			});
-			req.end();
-		});
-	}
 	/**
 	 * Makes a http/https get request to the given link and returns the status, headers, and a stream.
 	 * The request data itself can be read with ReadStream#read().
 	 * @param opts request opts - headers, etc.
+	 * @param body POST body
 	 */
-	async getFullResponse(opts?: AnyObject): Promise<{
-		statusCode: number | undefined, statusMessage: string | undefined,
-		headers: http.IncomingHttpHeaders | undefined, stream: Streams.ReadStream,
-	} | null> {
-		if (Config.noNetRequests) throw new Error(`Net requests are disabled.`);
-		return new Promise(resolve => {
+	getFullResponse(opts: NetRequestOptions = {}, body?: PostData): Promise<{
+		statusCode: number | undefined,
+		statusMessage: string | undefined,
+		headers: http.IncomingHttpHeaders | undefined,
+		stream: Streams.ReadStream,
+	}> {
+		if (opts.body) {
+			if (body) throw new Error(`must not pass both body and opts.body`);
+			body = opts.body;
+		}
+
+		if (body && typeof body !== 'string') {
+			let out = '';
+			for (const key in body) {
+				if (out) out += `&`;
+				out += `${key}=${encodeURIComponent('' + body[key])}`;
+			}
+			body = out;
+		}
+
+		const postBody = body;
+		if (postBody) {
+			if (!opts.headers) opts.headers = {};
+			opts.headers['Content-Type'] = 'application/x-www-form-urlencoded';
+			opts.headers['Content-Length'] = postBody.length;
```

Comment:
This will not handle Unicode properly:

```
bastille% node -e 'console.log("\u{1F914}".length)'
2
```

There are 4 bytes in this string, which is what the `Content-Length` header should include as a length if this were the request's body. You can get the proper length with `Buffer.byteLength(postBody)`.

Reply:
Thanks!

Reply:
I actually did think of that problem right after pushing this, but I didn't know `Buffer.byteLength` existed.